src/wallet/spend.cpp

```diff
@@ -601,34 +630,38 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, AvailableCoins
 
         // If possible, fund the transaction with confirmed UTXOs only. Prefer at least six
         // confirmations on outputs received from other wallets and only spend confirmed change.
-        if (auto r1{AttemptSelection(wallet, value_to_select, CoinEligibilityFilter(1, 6, 0), coins, coin_selection_params)}) return r1;
-        if (auto r2{AttemptSelection(wallet, value_to_select, CoinEligibilityFilter(1, 1, 0), coins, coin_selection_params)}) return r2;
+        // Also avoid mixing different UTXO OutputTypes to preserve privacy
+        if (auto r1{AttemptSelection(wallet, value_to_select, CoinEligibilityFilter(1, 6, 0), coins, coin_selection_params, /*allow_mixed_output_types=*/false)}) return r1;
+        if (auto r2{AttemptSelection(wallet, value_to_select, CoinEligibilityFilter(1, 1, 0), coins, coin_selection_params, /*allow_mixed_output_types=*/false)}) return r2;
+
+        // If we can't fund the tx from confirmed UTXOs of a single OutputType, allow mixed output types
+        if (auto r3{AttemptSelection(wallet, value_to_select, CoinEligibilityFilter(1, 1, 0), coins, coin_selection_params, /*allow_mixed_output_types=*/true)}) return r3;
```

Comment:
In df6aa1d597b4de1f4707719b3ad5518a94645ee1 "wallet: run coin selection by `OutputType`"

AFAICT, this duplicates the selection attempts with segregated types for `CoinEligibilityFilter(1, 1, 0).

Either way, the selection with segregated types is attempted first and mixing only happens as a last resort. So, directly setting `allow_mixed_output_types=true` should result in equivalent back-off steps without repeating the segregated output attempts before allowing mixing. 

Not adding this line also obsoletes the following commit to fix the duplicate appearance of `r3`,  321a13e177dc79581ec896568ab8e7633d95b17f "refactor: update variable names".
 
 ```suggestion
        // If we can't fund the tx from confirmed UTXOs of a single OutputType, allow mixed output types
        if (auto r2{AttemptSelection(wallet, value_to_select, CoinEligibilityFilter(1, 1, 0), coins, coin_selection_params, /*allow_mixed_output_types=*/true)}) return r2;
 ```

Reply:
ah, good catch! `AttemptSelection` will only try with mixed inputs if it is unable to find at least one solution from segregated inputs and if mixing is allowed. So yes, running `r2` with `allow_mixed_inputs=true` achieves want we want without needing the duplicated r3